lib/Dialect/EmitC/Conversion/TosaToEmitC.cpp

```diff
@@ -108,11 +108,149 @@ class RsqrtOpConversion : public OpConversionPattern<tosa::RsqrtOp> {
   }
 };
 
+template <typename SrcOp>
+SmallVector<Value, 2>
+create_broadcast_op_if_needed(SrcOp &srcOp, ArrayRef<Value> operands,
+                              ConversionPatternRewriter &rewriter) {
+  // TOSA allows implicit broadcasting, so we need to insert broadcast_in_dim
+  // ops if necessary, e.g.:
+  // tensor<8xf32>     -> tensor<1x4x4x8xf32> Broadcast dims = (3)
+  // tensor<4x8xf32>   -> tensor<1x4x4x8xf32> Broadcast dims = (2, 3)
+  // tensor<4x4x8xf32> -> tensor<1x4x4x8xf32> Broadcast dims = (1, 2, 3)
+
+  StringRef broadcastFuncName = "emitc::broadcast_in_dim";
+  StringAttr broadcastCallee = rewriter.getStringAttr(broadcastFuncName);
+
+  Value output = srcOp.getResult();
+  auto opOutputShape = output.getType().cast<RankedTensorType>().getShape();
+  auto opOutputRank = output.getType().cast<RankedTensorType>().getRank();
+  SmallVector<Value, 2> broadcastedOperands;
+  broadcastedOperands.push_back(operands[0]);
+  broadcastedOperands.push_back(operands[1]);
+  for (size_t i = 0; i < operands.size(); i++) {
+    auto &operand = operands[i];
+    auto operandShape = operand.getType().cast<RankedTensorType>().getShape();
+    auto operandRank = operand.getType().cast<RankedTensorType>().getRank();
+
+    // Insert a broadcast_in_dim Operation if shape of operands don't match
+    if (!operandShape.equals(opOutputShape)) {
+      SmallVector<Attribute, 1> broadcastIndices;
+      auto numBroadcastDims = opOutputRank - operandRank;
+      for (int64_t d = numBroadcastDims; d < opOutputRank; d++) {
+        broadcastIndices.push_back(
+            mlir::IntegerAttr::get(rewriter.getIntegerType(64), d));
+      }
+
+      RankedTensorType ty =
+          RankedTensorType::get({static_cast<int64_t>(operandRank)},
+                                IntegerType::get(srcOp.getContext(), 64));
+      ArrayAttr broadcastArgs = rewriter.getArrayAttr(
+          DenseIntElementsAttr::get(ty, broadcastIndices));
+
+      ArrayAttr templateBroadcastArgs =
+          rewriter.getArrayAttr({TypeAttr::get(srcOp.getType())});
+
+      auto broadcastArg = rewriter.create<emitc::CallOp>(
+          srcOp->getLoc(), srcOp.getType(), broadcastCallee, broadcastArgs,
+          templateBroadcastArgs, operand);
+      // Replace the original operand with the result of the broadcast_in_dim
+      // operation
+      broadcastedOperands[i] = broadcastArg.getResult(0);
+    }
+  }
+  return broadcastedOperands;
+}
+
+template <typename SrcOp>
+class CallOpBroadcastableConversion : public OpConversionPattern<SrcOp> {
```

Comment:
Maybe we should name this `BroadcastableCallOpConversion` instead of `CallOpBroadcastableConversion`? :)

@simon-camp @dajuro Your suggestions?

Reply:
No preference here.

Reply:
Let's just keep it as it. We can still refactor in a followup if desired.